adam/flask_backend/app.py

```diff
@@ -135,7 +147,133 @@ def get_scene() -> Any:
     }
 
 
+@app.route("/api/get_image", methods=["GET"])
+def get_image() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    image_path = request.args.get("image_path", default="").strip("/")
+    image_file = DATA_DIR.joinpath(image_path).resolve()
+    if not is_relative_to(image_file, DATA_DIR):
+        return {"message": "Directory out of range"}, HTTPStatus.BAD_REQUEST
+    with open(image_file, "rb") as fp:
+        image_data = b64encode(fp.read()).decode("utf-8")
+    return {"image_data": image_data}
+
+
+@app.route("/api/poll_changes", methods=["GET"])
+def poll_changes() -> Any:
+    if not request.args:
+        abort(HTTPStatus.BAD_REQUEST)
+    learner = request.args.get("learner", default="")
+    training_curriculum = request.args.get("training_curriculum", default="")
+    testing_curriculum = request.args.get("testing_curriculum", default="")
+    if not learner and not training_curriculum and not testing_curriculum:
+        abort(HTTPStatus.BAD_REQUEST)
+
+    if any("/" in arg for arg in (learner, training_curriculum, testing_curriculum)):
+        abort(HTTPStatus.BAD_REQUEST)
+
+    curriculum_dir = (
+        LEARNERS_DIR
+        / learner
+        / EXPERIMENTS_DIR_NAME
+        / training_curriculum
+        / EXPERIMENTS_TESTING_DIR_NAME
+        / testing_curriculum
+    ).resolve()
+
+    if not is_relative_to(curriculum_dir, LEARNERS_DIR):
+        return {"message": "Directory out of range"}, HTTPStatus.BAD_REQUEST
+
+    if (
+        app.config["LEARNER_DIR_CONTENTS"].get(learner) is None
+        or app.config["LEARNER_DIR_CONTENTS"][learner].get(training_curriculum) is None
+        or app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum].get(
+            testing_curriculum
+        )
+        is None
```

Comment:
Nit: Since we're not using the result of `.get()` here, It would be slightly clearer to write this using `not in`:

```python
learner not in app.config["LEARNER_DIR_CONTENTS"]
or training_curriculum not in app.config["LEARNER_DIR_CONTENTS"][learner]
or testing_curriculum not in app.config["LEARNER_DIR_CONTENTS"][learner][training_curriculum]
```